presto-main/src/test/java/com/facebook/presto/metadata/TestFunctionRegistry.java

```diff
@@ -136,6 +142,236 @@ public void testListingHiddenFunctions()
         assertFalse(names.contains("like"), "Expected function names " + names + " not to contain 'like'");
     }
 
+    @Test
+    public void testResolveFunctionByExactMatch()
+            throws Exception
+    {
+        assertThatResolveFunction()
+                .among(functionSignature("bigint", "bigint"))
+                .forParameters("bigint", "bigint")
+                .returns(functionSignature("bigint", "bigint"));
+    }
+
+    @Test
+    public void testResolveTypeParametrizedFunction()
+            throws Exception
+    {
+        assertThatResolveFunction()
+                .among(functionSignature(ImmutableList.of("T", "T"), ImmutableList.of(typeVariable("T"))))
+                .forParameters("bigint", "bigint")
+                .returns(functionSignature("bigint", "bigint"));
+    }
+
+    @Test
+    public void testResolveFunctionWithCoercion()
+            throws Exception
+    {
+        assertThatResolveFunction()
+                .among(
+                        functionSignature("decimal(p,s)", "double"),
+                        functionSignature("decimal(p,s)", "decimal(p,s)"),
+                        functionSignature("double", "double")
+                )
+                .forParameters("bigint", "bigint")
+                .returns(functionSignature("decimal(19,0)", "decimal(19,0)"));
+    }
+
+    @Test
+    public void testAmbiguousCallWithNoCoercion()
+            throws Exception
+    {
+        assertThatResolveFunction()
+                .among(
+                        functionSignature("decimal(p,s)", "decimal(p,s)"),
+                        functionSignature(ImmutableList.of("T", "T"), ImmutableList.of(typeVariable("T")))
+                )
+                .forParameters("decimal(3,1)", "decimal(3,1)")
+                .returns(functionSignature("decimal(3,1)", "decimal(3,1)"));
```

Comment:
This particular tests is intended to verify that among 2 function (generic and specific) that match exactly for the input types, result function is being selected. According to the algorithm we have implemented, the first one (`functionSignature("decimal(p,s)", "decimal(p,s)")`) must be chosen. I will add some extra check here to make it less confusing.